02-peer-protocol.md

```diff
@@ -430,6 +446,515 @@ would create a Denial of Service risk; therefore, forgetting it is recommended
+  - MUST set `witness` to the serialized witness data for the input
+    corresponding to `prevtx_txid`:`prevtx_vout`
+  - MUST remember the details of this funding transaction.
+  - MUST NOT send a `witness_stack` whose length exceeds its
+    previously indicated `max_witness_len`.
+  - if it has NOT received a valid `commitment_signed` message:
+    - MUST NOT send a `funding_signed2` message.
+
+The receiving node:
+  - if the `witness_stack` length exceeds `max_witness_len`:
+    - MUST error.
+  - MUST verify it has received valid commitment signatures
+  - SHOULD apply `witness`es to the funding transaction and
+    broadcast it.
+
+#### Rationale
+
+Every node must transmit their complete set of witness data for
+the inputs included in the funding transaction.
+
+`prevtx_txid` and `prevtx_vout` are the identifier for the input
+which the witness data pertain to
+
+`witness` is the data for a witness element in a witness stack
+
+### Kicking Off Replace-By-Fee: `init_rbf`
+
+This message initiates the flow to create a replacement funding
+transaction.  It is sent by the channel opener, after the funding
+transaction has been broadcast but before a `funding_locked` message
+is exchanged.
+
+Once an `init_rbf` message has been successfully ack'd by the accepter
+node, the message flow returns to `commitment_signed` and proceeds as
+indicated above.
+
+1. type: 62 (`init_rbf`)
+2. data:
+    * [`channel_id`:`channel_id`]
+    * [`u64`:`funding_satoshis`]
+    * [`u32`:`feerate_per_kw`]
+    * [`u32`:`feerate_per_kw_funding`]
+    * [`u16`:`num_additional_inputs`]
+    * [`num_additional_inputs*input_info`:`input_info`]
+    * [`u16`:`num_outputs`]
+    * [`num_outputs*output_info`:`output_info`]
+
+#### Requirements
+
+The sending node:
+  - MUST have sent `open_channel2`.
+  - MUST send a `feerate_per_kw_funding` greater than the most recently
+    negotiated rate.
+  - MAY update the `feerate_per_kw`, the commitment transaction feerate.
+  - MAY include additional inputs.
+  - MAY set the `num_inputs` to zero.
+  - MUST transmit all outputs, excluduing the channel funding output.
+  - MAY include at least one output with the `sats` set to zero,
+    to be used as the change output.
+
+The receiving node:
+  - MUST return an error if:
+    - the `feerate_per_kw_funding` is not greater than the previously
+      negotiated rate.
+  - MUST return an error, but not fail the channel if:
+    - it has already received a `funding_locked` message
+
+
+#### Rationale
+The sending node's half of the funding transaction will be composed with
+all of the previously relayed inputs plus the ones included here.
+
+All outputs must be transmitted in the `init_rbf` message, including
+ones communicated previously, except for the channel funding output.
+
+If a valid `funding_locked` message is received in the middle of an
+RBF workflow, the RBF attempt MUST be abandoned.
+
+
+### Acknowledging Replace-By-Fee `ack_rbf`
```

Comment:
What problem the ack_rbf is handling ? Can't receiver of init_rbf directly replies with a commitment_signed ?

Reply:
good point, i think it is infact superfluous.

Reply:
update: in order to avoid reversing the order of `commitment_signed` exchange, you need to get at least one message back from the peer before you send them the `commitment_signed` message.

Reply:
Can't the `init_rbf` sender joins a `commitment_signed` and interpret the lack of getting back a `commitment_signed` as a refusal (though harder for the state machine I guess) ?